### PR TITLE
✨ [portfolio] Add keyword filter & update UI

### PR DESCRIPTION
**Pull request by AuroraHuang22**
截圖 2023-04-25 上午12 39 12
截圖 2023-04-25 上午12 44 56

### CODE REVIEW

File: src/store/modules/nft.js [hunk 1]
```diff
@@ -493,12 +518,21 @@ const actions = {
...
+    ];
+    const mergedClassIds = [...new Set([...currentClassIds, ...nftClassIds])];
     await Promise.all(
-      nftClassIds.map(classId => {
+      mergedClassIds.map(classId => {
         const promises = [
           catchAxiosError(
             dispatch('populateNFTClassMetadataFromURIAndISCN', classId)
           ),
+          catchAxiosError(dispatch('fetchIscnRecords', classId)),
```
Comment (williamchong) [inline src/store/modules/nft.js:535]: A potential optimization is to first get the iscnIds from all nft classes, then deduplicate the iscnIds.
An iscnId can be parent of multiple nft classes

File: src/store/modules/nft.js [hunk 2]
```diff
@@ -493,12 +518,21 @@ const actions = {
     });
 
     const nftClassIds = Array.from(nftClassIdDataMap.keys());
+    const currentClassIds = [
```
Comment (williamchong) [inline src/store/modules/nft.js:521]: How does this dffer from nftClassIds?

File: src/pages/dashboard/index.vue [hunk 3]
```diff
@@ -63,11 +63,15 @@
...
         :is-loading-portfolio-items="isLoading"
+        :nft-keyword-list="nftKeywordList"
+        :nft-keyword-filtering="nftKeywordsFilter"
         @portfolio-change-tab="handleTabChange"
         @portfolio-change-sorting="handleNFTClassListSortingChange"
         @portfolio-change-creator="handleNFTClassListCreatorChange"
         @portfolio-change-type="handleNFTClassListTypeChange"
+        @portfolio-change-keywords="handleNFTKeywordsChange"
         @infinite-scroll="handleInfiniteScroll"
+        @portfolio-reset-filter="handleCleanFilter"
```
Comment (williamchong) [inline src/pages/dashboard/index.vue:74]:
```suggestion
        @portfolio-reset-filter="handleClearFilter"
```

File: src/mixins/portfolio.js [hunk 4]
```diff
@@ -576,6 +624,34 @@ export const createPortfolioMixin = ({
...
+        logTrackerEvent(
+          this,
+          'portfolio',
+          `portfolio_filter_keywords`,
+          `Filter portfolio item by ${value}`,
+          1
+        );
+      }
+    },
+    handleCleanFilter() {
```
Comment (williamchong) [inline src/mixins/portfolio.js:649]:
```suggestion
    handleClearFilter() {
```

File: src/mixins/portfolio.js [hunk 5]
```diff
@@ -576,6 +624,34 @@ export const createPortfolioMixin = ({
...
+      this.resetFilter = true;
+      this.nftKeywordsFilter = [];
+      this.nftCreatorFilter = [];
+      this.resetFilter = false;
```
Comment (williamchong) [inline src/mixins/portfolio.js:653]: This is a sync function, just set `resetFilter` false?

File: src/locales/en.json [hunk 6]
```diff
@@ -386,7 +386,11 @@
   "order_menu_by": " By {order}",
   "order_menu_collected": "Own",
   "order_menu_display_state": "Featured",
-  "order_menu_filter_by": "Filter by",
+  "order_menu_filter_by": "Filter",
+  "order_menu_filter_by_creators": "Creators",
+  "order_menu_filter_by_keywords": "Keywords",
+  "order_menu_filter_by_type": "NFT type",
+  "order_menu_filter_clean": "Clean Filter",
```
Comment (williamchong) [inline src/locales/en.json:393]:
```suggestion
  "order_menu_filter_clean": "Clear Filter",

```

File: src/components/NFTPortfolio/MainView.vue [hunk 7]
```diff
@@ -38,128 +38,69 @@
...
+
+            :portfolio-items-type-filtering="portfolioItemsTypeFiltering"
+            :portfolio-collected-creator-list-with-sorting="portfolioCollectedCreatorListWithSorting"
+            :nft-keyword-list="nftKeywordList"
+            :nft-keyword-filtering="nftKeywordFiltering"
+
+            @filter-click-type-filter="selectedFilter = 'type'"
+            @filter-click-creator-filter="selectedFilter = 'creators'"
+            @filter-click-keyword-filter="selectedFilter = 'keywords'"
+            @filter-click-clean-filter="handleResetFilter"
```
Comment (williamchong) [inline src/components/NFTPortfolio/MainView.vue:66]:
```suggestion
            @filter-click-clear-filter="handleResetFilter"
```

File: src/components/NFTPortfolio/FilterDropdown.vue [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+      this.$emit('filter-click-type-filter');
+    },
+    handleClickCreators() {
+      this.$emit('filter-click-creator-filter');
+    },
+    handleClickKeywords() {
+      this.$emit('filter-click-keyword-filter');
+    },
+    handleResetFilter() {
+      this.$emit('filter-click-clean-filter');
```
Comment (williamchong) [inline src/components/NFTPortfolio/FilterDropdown.vue:170]:
```suggestion
      this.$emit('filter-click-clear-filter');
```

**Review by williamchong — CHANGES_REQUESTED** [review]

File: src/store/modules/nft.js (hunk 1 above)
Comment (AuroraHuang22) [reply src/store/modules/nft.js:535]: I merged `fetchIscnRecords` into `populateNFTClassMetadataFromURIAndISCN`.

[the remark this answers is not in the document]
Comment (AuroraHuang22) [reply src/store/modules/nft.js:535]: I merged `fetchIscnRecords` into `populateNFTClassMetadataFromURIAndISCN`.

File: src/mixins/portfolio.js (hunk 5 above)
Comment (williamchong) [reply src/mixins/portfolio.js]: @AuroraHuang22 still not sure what is the logic behind
```
      this.resetFilter = true;
      ... (sync functions)
      this.resetFilter = false;
```

**Review by williamchong — APPROVED** [review]

**Review by nwingt — APPROVED** [review]

**Pull request merged**